x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/allocation/TrainedModelAllocationNodeService.java

```diff
@@ -0,0 +1,384 @@
+                // kicks off asynchronous cluster state update
+                handleLoadSuccess(deployedTask);
+            } catch (Exception ex) {
+                // kicks off asynchronous cluster state update
+                handleLoadFailure(loadingTask, ex);
+            }
+        }
+    }
+
+    public void stopDeploymentAndNotify(TrainedModelDeploymentTask task) {
+        ActionListener<Void> notifyDeploymentOfStopped = ActionListener.wrap(
+            stopped -> updateStoredState(
+                task.getModelId(),
+                new RoutingStateAndReason(RoutingState.STOPPED, ""),
+                ActionListener.wrap(s -> {}, failure -> {})
+            ),
+            failed -> { // if we failed to stop the process, something strange is going on, but we should still notify of stop
+                logger.warn(() -> new ParameterizedMessage("[{}] failed to stop due to error", task.getModelId()), failed);
+                updateStoredState(
+                    task.getModelId(),
+                    new RoutingStateAndReason(RoutingState.STOPPED, ""),
+                    ActionListener.wrap(s -> {}, failure -> {})
+                );
+            }
+        );
+        updateStoredState(
+            task.getModelId(),
+            new RoutingStateAndReason(RoutingState.STOPPING, "task locally canceled"),
+            ActionListener.wrap(success -> stopDeploymentAsync(task, notifyDeploymentOfStopped), e -> {
+                if (ExceptionsHelper.unwrapCause(e) instanceof ResourceNotFoundException) {
+                    logger.debug(
+                        () -> new ParameterizedMessage(
+                            "[{}] failed to set routing state to stopping as allocation already removed",
+                            task.getModelId()
+                        ),
+                        e
+                    );
+                } else {
+                    // this is an unexpected error
+                    // TODO this means requests may still be routed here, should we not stop deployment?
+                    logger.warn(
+                        () -> new ParameterizedMessage("[{}] failed to set routing state to stopping due to error", task.getModelId()),
+                        e
+                    );
+                }
+                stopDeploymentAsync(task, notifyDeploymentOfStopped);
+            })
+        );
+    }
+
+    public void infer(TrainedModelDeploymentTask task, String input, TimeValue timeout, ActionListener<InferenceResults> listener) {
+        deploymentManager.infer(task, input, timeout, listener);
+    }
+
+    private TaskAwareRequest taskAwareRequest(StartTrainedModelDeploymentAction.TaskParams params) {
+        final TrainedModelAllocationNodeService trainedModelAllocationNodeService = this;
+        return new TaskAwareRequest() {
+            final TaskId parentTaskId = new TaskId(nodeId, taskIdGenerator.incrementAndGet());
```

Comment:
What is the parent task? When not use `TaskId#EMPTY_TASK_ID` as suggested in the comment in `TaskAwareRequest.java`